drivers: Add interrupt controller API

bjarki-andreasen opened the pull request:
This PR has now been updated to implement the solution described in this issue https://github.com/zephyrproject-rtos/zephyr/issues/67583

Size comparisons:
```
                        IRQ             SHARED          DYNAMIC         LVL        LTO
IRQ FLASH           38244 B            39652 B          39648 B     39776 B    34436 B
IRQ RAM              5912 B             5912 B           7848 B      7848 B     7840 B
SYS IRQ FLASH       42276 B            42276 B          42676 B     42676 B    36776 B
SYS IRQ RAM          5912 B             5912 B           6104 B      6104 B     5912 B
```
(note that no dynamic irq is registered, so LTO optimizes out the dynamic irq list)

Some example output of the generated code :)
Generated handlers:
```
static inline void __sys_intl_handler_8_5(void)
{
	/*
	 * Interrupt generating device: /soc/peripheral@50000000/clock@5000
	 */
	if (__sys_irq_handler_8_5_75()) {
		return;
	}

	/*
	 * Interrupt generating device: /soc/peripheral@50000000/power@5000
	 */
	if (__sys_irq_handler_8_5_117()) {
		return;
	}

	if (sys_irq_dynamic_handler(1)) {
		return;
	}

	sys_irq_log_spurious_intl(8, 5);

	sys_irq_spurious_handler();
}
```
Generated IRQ enumerations:
```
/*
 * Interrupt controller: /soc/interrupt-controller@e000e100
 * Interrupt line: 3
 * Interrupt generating device: /soc/spu@50003000
 */
#define SYS_DT_IRQN_8_3 0

/*
 * Interrupt controller: /soc/interrupt-controller@e000e100
 * Interrupt line: 5
 * Interrupt generating device: /soc/peripheral@50000000/clock@5000
 * Interrupt generating device: /soc/peripheral@50000000/power@5000
 */
#define SYS_DT_IRQN_8_5 1
```

A very cool solution I want to note as well is this macro:
```
INTC_DT_DEFINE_IRQ_VECTOR(node_id)
{
        return 1;
}
```
which wraps the existing ARCH_ISR_DIRECT_DECLARE(name) macro synthesizing the name of the function to match the weakly defined default vector in the vector table, allowing the existing infrastructure to be used, including ISR_DIRECT_PM()

## For arch review
This PR introduces two new requirements:
* We need to tell the kernel to generate interrupt handlers for devices which don't have status "okay" to:
  * Generate interrupt handlers for reserved devices, like RTCs reserved for sys tick or other SoC specific features
  * Generate interrupt handlers for devices, but not enable a driver for unit testing.
* Interrupt handlers need to signal whether they handled an IRQ or not in case of:
  * Shared IRQs to inform the kernel when we can stop walking IRQ handlers
  * Shared and normal IRQs, to allow the kernel to detect spurious interrupts

To generate interrupt handlers, but exclude device drivers, the status "reserved" is used to mean "supported by kernel, don't include a driver for it".

IRQ handlers will signal whether they handled the IRQ or not by returning 1 if handled, 0 if not handled.

## Analyzing code

**INTC_DT_DEFINE_IRQ_VECTOR with and without LTO**
```
mov     r0, sp
bic.w   r1, r0, #7
mov     sp, r1
push    {r0, lr}
ldr     r0, [pc, #16]
bl      0x2b00 <k_sem_give> /* first call in direct isr body (irq handler) */
```

**SYS_DT_DEFINE_IRQ_HANDLER without LTO**
```
mov     r0, sp
bic.w   r1, r0, #7
mov     sp, r1
push    {r0, lr}
bl      75c <__sys_irq_handler_7_20_107>
push    {r0, r1, r2, lr}
ldr     r3, [pc, #32]   ; (780 <__sys_irq_handler_7_20_107+0x24>)
ldr     r2, [pc, #32]   ; (784 <__sys_irq_handler_7_20_107+0x28>)
str     r3, [sp, #4]
ldr     r3, [pc, #32]   ; (788 <__sys_irq_handler_7_20_107+0x2c>)
ldr     r1, [pc, #36]   ; (78c <__sys_irq_handler_7_20_107+0x30>)
str     r3, [sp, #0]
movs    r0, #1
movs    r3, #34 ; 0x22
bl      158 <z_zassert.constprop.0>
```

**SYS_DT_DEFINE_IRQ_HANDLER with LTO**
```
mov     r0, sp
bic.w   r1, r0, #7
mov     sp, r1
push    {r0, lr}
ldr     r3, [pc, #40]   ; (3f14 <__intc_vector_7_20+0x34>)
sub     sp, #8
str     r3, [sp, #4]
ldr     r3, [pc, #36]   ; (3f18 <__intc_vector_7_20+0x38>)
ldr     r2, [pc, #40]   ; (3f1c <__intc_vector_7_20+0x3c>)
ldr     r1, [pc, #40]   ; (3f20 <__intc_vector_7_20+0x40>)
str     r3, [sp, #0]
movs    r0, #1
movs    r3, #34 ; 0x22
bl      581c <z_zassert.constprop.0>
```

**ISR_DIRECT_DECLARE without LTO**
```
mov     r0, sp
bic.w   r1, r0, #7
mov     sp, r1
push    {r0, lr}
ldr     r0, [pc, #16]   ; (1d0 <CONFIG_SRAM_SIZE+0x10>)
bl      4810 <z_impl_k_sem_give>
```

**ISR_CONNECT with and without LTO**
```
push    {r3, lr}
mrs     r3, IPSR
ldr     r2, [pc, #24]   ; (1500 <_isr_wrapper+0x20>)
subs    r3, #16
add.w   r1, r2, r3, lsl #3
ldr.w   r0, [r2, r3, lsl #3]
ldr     r1, [r1, #4]
blx     r1
```

carlescufi commented:
Arch WG meeting:

- @bjarki-trackunit presents the proposal
- @carlescufi asks if the current `irq_enable()`-like APIs will be completely replaced by `intc_` ones. @bjarki-trackunit answers that yes, they will
- @gmarull asks if the compile-time checks will remain. @bjarki-trackunit says yes, with the help of the single-linked list
- @carlescufi mentions direct ISRs and the current interrupt model with the software ISR table
- @nashif mentions that the IRQ code needs to be available very early on (even pre-C)
- @gmarull asks about the Devicetree layout. One of the issues is about GPIO being both a GPIO controller and an interrupt controller. @bjarki-trackunit explains how he's solved it
-

andyross commented on drivers/intc/intc_arm_v8m_nvic.c:
```diff
@@ -0,0 +1,223 @@
+		if (irq != NULL) {								\
+			irq->handler(irq);							\
+		} else {									\
+			while (1) {								\
```
Not understanding this at all.  Presumably you get here on a spurious interrupt?  We have an existing spurious interrupt handler API the arch layers all honor, as that's a pretty routine glitch during device bringup (e.g. "oops typo in the interrupt number", or "oops, the bootloader left the UART RX interrupt enabled" -- absolutely routine things).  Entering a silent infinite loop seems like exactly the wrong thing to do.

andyross commented on include/zephyr/drivers/intc.h:
```diff
@@ -0,0 +1,462 @@
+
+/** Interrupt request handle */
+struct intc_irq {
+	/** Interrupt request single-linked list node */
```
Actually, it doesn't go anywhere?  I can't find code that uses this.  Is this a forgotten attempt to implement shared interrupts?  As of right now it doesn't seem to work?

bjarki-trackunit replied:
Its supposed to be used for shared IRQs, and for secondary interrupt controllers like GPIOs, which usually only store a few IRQs relative to how many pins they have.

andyross reviewed (commented):
So... I don't hate the API, it's clean and reasonable (though it's also wordy, requiring three steps to do what a single IRQ_CONNECT() does right now).

But this seems like only about 1/3 of the solution.  This is all about the user/driver/subsystem-facing API.  But the API we have for that is really... not so bad right now.  What's the story for why we want to merge this?  What can we do with it that we can't right now?

The problems with our interrupt layer (and it has problems!) are on the other side:

* There's no management of interrupt controller cascade by the framework.  Every platform needs to write its own glue to call child interrupt controller handlers out of their parents, even though the structure of that tree is known at build time by devicetree.  This PR seems targetted at only NVIC platforms without a secondary controller, meaning that it's really not seeing the mess we're currently living with.

* The interrupt ID/number/index/offset management (i.e. the mapping between a given device interrupt and the index into the _sw_isr_table[] array) is likewise ad-hoc and platform dependent, and is generally implemented with a bunch of crazy magic numbers stored in kconfig.  And again, a clean mapping could be trivially derived from the graph of controllers known to DTS.

Also, it's worth pointing out that this scheme is throwing out efficiency features we've had for a long time: Interrupts are 100% runtime now, you have to write and call C code to get an interrupt registered, spending bytes on code that has to execute and RAM on data structures that have to be mutated.  Some platforms even today are still RAM-constrained and would like their ISR tables in pre-built flash.  Maybe we don't care about apps like that anymore, it's an argument worth having.  But we should be clear about what we're walking away from.

tl;dr: I guess my feeling is that if we're going to rewrite interrupt handling, we should sharpen the machetes and do it all, for everyone.

andyross commented:
Looks like the bot didn't assign reviewers, maybe because it's a draft?  For sure @dcpleung @peter-mitsis @carlocaione @npitre want to get in on this.  @tbursztyka should expand my note about MSI.   Lots of platform folks will have important input too, especially ones who've had to deal with cascaded controllers.

bjarki-trackunit commented:
@andyross This API aims to solve points 1. and 2. while adding support for "secondary" interrupt controllers, like GPIO ports and other IO expanders which can route interrupts through IO pins.

1. Interrupt cascading
This API specifies every IRQ as an interrupt line of a specific interrupt controller, like a GPIO pin is a pin of a specific GPIO port. This reflects the devicetree (hardware) layout completely, as opposed to the current "absolute" ISR table, which tries to flatten the layout into a single list, which has spawned the need for encoding the IRQ number in levels.

2. IRQ encoding
This solution does not require IRQ encoding, as each IRQ number is relative to an the IRQs interrupt parent. Each parent manages its own list of IRQs in whicever way is most efficient. There is no absolute ISR table using this API, there can be as many tables as there are interrupt controllers, some interrupt controllers (like GPIOs) may simply store the few IRQs they manage in a single linked list to save memory as they do now.

3. Efficiency
This solution does remove the possibility of storing the SW IRQ table in ROM, but it can be quite a bit more efficient at storing IRQs, given that GPIOs may store their IRQs in single linked lists as they do now rather than an array of however many pins they have, for a chip with 4 ports of 32 pins, that'quite a bit.

For really ressource constrained devices, direct IRQs can be used, without the need for creating a SW IRQ table. Direct IRQs can be created for every interrupt controller at any level with this API, including a GPIO pin :)

rakons commented on include/zephyr/drivers/intc.h:
```diff
@@ -0,0 +1,462 @@
+ *
+ * Expands to __intc_irq_handler_<ord>_<irq>
+ */
+#define INTC_IRQ_HANDLER_NAME(ord, irq) \
```
Is this meant to be used anywhere else than in the macro below? I am asking because it requires very strict values here and it seems little risky to reveal the macro as an part of the API.

bjarki-trackunit replied:
Its meant to be used within device drivers to populate their interrupt vector tables, maybe it can be moved to a less user facing header like direct_irq_handler.h :)

bjarki-trackunit commented on include/zephyr/drivers/intc.h:
```diff
@@ -0,0 +1,462 @@
+/** @cond INTERNAL_HIDDEN */
+
+/** Interrupt request specification */
+struct intc_irq_spec {
```
The issue with the old-style number is that while being simple, it is lacking vital information, namely which interrupt controller the interrupt belongs to. The old style number is therefore only reflective of the devicetree, allowing `DT_IRQN` to return the correct IRQ number, if there is only one interrupt controller, at which point the instance can be omitted, and the API can just be static (as it is now).

An interrupt number is the interrupt line number of a specific interrupt controller. To clearly visualize the issue, consider how you would rewrite the GPIO API to only use an integer for the pin number, with a single API.

andyross replied:
> it is lacking vital information

It's not.  Or rather it doesn't have to be.  The controller is extractable if you know how the table is packed, which right now requires interpreting a bunch of kconfig values storing sizes and offets, and it's a mess.  Which is why I'd like to see it reworked.

But that's not an indictment of using an integer.  We can totally continue to use integers to identify interrupts.  And (implemented correctly) integers have some excellent efficiency benefits which would be good to preserve.

(Also, just to repeat: this API requires a separate step to construct and extract that spec struct, vs. just using a by-value macro like DT_IRQN().  That right off the bat makes this a **worse** API, not a better one.  It's not a critical problem, but it's still a cost we're paying for exposing the spec abstraction to the user)

bjarki-trackunit replied:
Can you see the parallel I'm drawing between the GPIO API and the IRQ API? We have no problem pairing a GPIO controller to a pin and passing this to the GPIO API (in fact, its rather simple and elegant IMO), why treat IRQs differently?

bjarki-trackunit commented on include/zephyr/drivers/intc.h:
```diff
@@ -0,0 +1,462 @@
+	/** Interrupt request priority */
+	uint16_t priority;
+	/** Interrupt request flags */
+	uint32_t flags;
```
the priority and irq number are integers, what will probably never exceed an 65365, but do exceed 255, where the flags is a bitfield, so it should be as large as we can provide :)

That said, I may update flags to be a size_t, but it does seem MSI fields can fit inside a uint32 https://i.stack.imgur.com/RwenJ.jpg

tbursztyka replied:
Seems like this is fine for MSI.

Anyway, the implementation of MSI/MSI-X is currently very much arch specific, that is: x86 specific. So there won't be any collision with such generic intc API. 

It could change someday, since interrupt management implementation in x86 is messy to say the least (you always need to allocate an irq to get a vector for instance, that does not make much sense in case of MSI-X or multi-vector MSI).

bjarki-trackunit commented on include/zephyr/drivers/intc.h:
```diff
@@ -0,0 +1,462 @@
+
+/** Interrupt controller API structure */
+struct intc_driver_api {
+	intc_api_request_irq request_irq;
```
I think it looks good, it may be more lines, but every API in the struct does not have to be split into multiple lines :)

bjarki-trackunit commented on include/zephyr/drivers/intc.h:
```diff
@@ -0,0 +1,462 @@
+ * @note Identical to INTC_DT_GET_IRQ_SPEC_BY_NAME() if IRQ exists
+ * @see INTC_DT_GET_IRQ_SPEC_BY_NAME()
+ */
+#define INTC_DT_GET_OPT_IRQ_SPEC_BY_NAME(node_id, name)				\
```
bus attached devices like sensors and RTCs are the rational for this API. Sensors often have 2-3 GPIO routed optional APIs, and there may be more than one instance of them, making `IS_ENABLED` and similar quite complex. See https://github.com/zephyrproject-rtos/zephyr/blob/30c660162f5fe70606bd122b61818ace47e6d908/drivers/sensor/lis2dh/lis2dh.c#L515-L526 for example (and note that this API is intended to replace the interrupt part of the GPIO API)

rakons commented:
I am not sure if I get the implementation correctly. What I get from here, is that it is trying to solve the issue with multi level interrupts and how they are implemented now. I cannot quite get if does it replace the whole current irq implementation or only for the nested controllers?
If it is planned to be used as a general API for top level interrupts I can see 2 main issues here:
1. Breaking the API. This gives us not only a lot of work on the Zephyr level, but also all the drivers, here and in external modules, would have to be updated.
2. Lack of the possibility to put SWI in ROM - for some of our smaller chips I can see it as a quite challenging to fit everything we need in RAM when we include SWI table here. I know that now we have some samples where a lot of work is put into cutting the RAM or ROM usage.

And as we could live with the need of API change, lacking the possibility to use ROM for SWI table may be serious drawback for some bigger samples.

For us (but please note it would be extremely biased) - only pros I can see are:
- The script to generate SWI tables is no longer required. So one step that can make troubles less.
- It does not block LTO usage - but this one we can have with the same cost of switching from ROM to RAM by the usage of dynamic interrupts, without the cost of API break.

It does not seem to conflict with #66392. At least not in a code level. But at the API level it might be an issue. I suppose, that we could optimize how nested interrupt controllers are handled using (linker) similar approach like in mentioned PR, but still encoding the nested interrupt in interrupt number. But I am also not 100% sure if this is the main goal of this PR.

@bjarki-trackunit - do you see any problem with merging both of this PRs? On code level, I am not even thinking there would be any conflict reported by git. But from conceptual point of view they seem to be totally different beasts.

bjarki-andreasen committed:
kernel: Add system IRQ iface, impl, script and cmake target

This commit adds the interface for the system IRQ subsystem,
which is split into three files:
- include/zephyr/sys/irq.h contains the user facing APIs and
  macros, which includes creating interrupt handlers, and
  enabling/configuring/disabling interrupts.
- include/zephyr/sys/irq_handler.h contains the interrupt
  line handlers interrupt controllers will call.
- include/zephyr/sys/internal/irq.h contains internal
  structures required by irq.c

The headers all require snippets generated by the script
gen_sys_irq.py included with this commit. This commit adds a
target and custom command to the top CMakeLists.txt to
ensure snippets are generated if SYS_IRQ=y.

Lastly, the source implementing the APIs provided with the
interface, irq.c, is added.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

bjarki-andreasen committed:
drivers: intc: Add intc device driver API header

Add intc.h device driver API header and inital Kconfig,
CMakeLists.txt and driver folder.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

bjarki-andreasen committed:
drivers: intc: Add arm v8m nvic device driver

Add arm v8m nvic device driver implementing the new intc.h
device driver API and interfacing with the new system IRQ
subsystem.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

bjarki-andreasen committed:
dts: interrupt-contoller: Add interrupt-lines base property

Add interrupt-lines property to interrupt-controller.yaml

This value will initially be optional, but it is required when
using the new system IRQ subsystem.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

bjarki-andreasen committed:
drivers: intc: Add irq_vector.h

Add irq_vector.h interface which wraps the existing
ISR_DIRECT_DECLARE() macro adapting it to the system IRQ schema.

ISR_DIRECT_DECLARE(name)
{}

is wrapped within this macro

INTC_DT_DEFINE_IRQ_VECTOR(node_id)
{}

which uses the node identifier to synthesize the name
of the declared function, strongly defining the weak
default vector in the vector table.

This allows for the existing PM macros, swap return value etc.
to be reused with the system IRQ schema.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

bjarki-andreasen committed:
arch: arm: irq.h: Use UTIL_CAT instead of ##

With the system IRQ INTC_DT_DEFINE_IRQ_VECTOR wrapper, the
argument must be unwrapped before being used in the function.

Using UTIL_CAT is therefore required. (it also looks nicer)

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

bjarki-trackunit commented:
## Notes
### Fully implemented shared and dynamic interrupts
To properly support shared and dynamic interrupts, a return value has been introduced to the IRQ handlers. If an IRQ handler handled the IRQ, it returns 1, otherwise 0. An example interrupt line handler follows:
```
static inline void __sys_intl_handler_8_5(void)
{
	/*
	 * Interrupt generating device: /soc/peripheral@50000000/clock@5000
	 */
	if (__sys_irq_handler_8_5_75()) {
		return;
	}

	/*
	 * Interrupt generating device: /soc/peripheral@50000000/power@5000
	 */
	if (__sys_irq_handler_8_5_117()) {
		return;
	}

	if (sys_irq_dynamic_handler(1)) {
		return;
	}

	sys_irq_log_spurious_intl(8, 5);

	sys_irq_spurious_handler();
}
```
The static IRQ handlers are attempted first, if none of them where the originator of the interrupt, the list of dynamic interrupts is iterated. If no registered or linked IRQ handler could handle the IRQ, the kernel will invoke a spurious IRQ fault.

### Spurious IRQ logging added
Spurious interrupts now produce the following output:
```
[00:00:00.411,224] <err> os: >>> ZEPHYR SPURIOUS IRQ: INTC ord: 8, INT line: 21
[00:00:00.411,285] <err> os: >>> ZEPHYR FATAL ERROR 1: Unhandled interrupt on CPU 0
[00:00:00.411,315] <err> os: Current thread: 0x200006a8 (unknown)
[00:00:00.434,722] <err> os: Halting system
```
This error is "kernel level", meaning it is generic across all architectures, interrupt controllers etc. Simply look to the devicetree_generated.h to find the problematic interrupt controller by devicetree ordinal, them follow the irq number :)

bjarki-andreasen committed:
tests: kernel: Add sys_irq test suites

Add test suites to validate the sys_irq API, shared IRQs,
dynamic IRQs and direct IRQs.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

pdgendt commented on drivers/intc/intc_arm_v8m_nvic/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+zephyr_library()
+
+zephyr_library_sources(arm_v8m_nvic.c)
+zephyr_linker_sources(ROM_START SORT_KEY 0x0vectors irq-vector-table.ld)
```
Will this be an issue for the `CMAKE_LINKER_GENERATOR`?

bjarki-andreasen replied:
I think so, I will need to look into making it more generic, it is a bit of a hack currently :)

bjarki-andreasen-nordic committed:
drivers: timer: nrf_rtc_timer: Update to system IRQ

Update nrf_rtc_timer.c to use system IRQ. This includes updating
the rtc1 nodes to be reserved rather than disabled.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
drivers: intc: add arm nvic v7m device driver

Add ARM NVIC V7M device driver

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
drivers: clock_control: update clock_control_nrf to sys IRQ

Update clock_control_nrf.c device driver to use system IRQ

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
drivers: serial: uart_nrfx_uarte.c: Update to use sys IRQ

Update uart_nrfx_uarte.c to use system IRQ

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
drivers: gpio: gpio_nrfx: Update to use sys IRQ

Update gpio_nrfx.c to use system IRQ

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
drivers: add power and driver for nrf_power

Add power subsystem and power driver for nrf_power using system IRQ.

TODO: Update nrfxlib to return handled or not, otherwise this or
clock isr will never be called if both are implemented.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen commented:
Hope to continue this within a couple of weeks

bjarki-andreasen commented:
## Update

* Fixed issue with LTO causing spurious interrupts (LTO optimized away entire vector table, fixed with __used attr)
* Cleaned up sys_irq test suites
* Rebased on latest main

bjarki-andreasen-nordic committed:
dts: arm: nordic: add interrupt-lines nrf5340

Add interrupt-lines properties to the nrf5340 cpuapp and cpunet.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen commented:
## Update
* Added test suite for direct IRQs
* Cleaned up a bit

bjarki-andreasen commented:
## Update

* Added `INTC_DT_FOREACH_INTL` and `INTC_DT_FOREACH_IMPL_INTL` macros for iterating over interrupt lines by interrupt controller. `INTC_DT_FOREACH_INTL` iterates over every interrupt line, whether any IRQ is implemented or not, while `INTC_DT_FOREACH_IMPL_INTL` only iterates over interrupt lines which are implemented (connected to node with status okay or reserved)
* Updated NVIC drivers to use the new foreach macros

bjarki-andreasen commented:
## Update

* Quite a bit of cleanup
* Added macros for getting the number of implemented and connected interrupt lines for each interrupt controller

bjarki-andreasen-nordic committed:
doc: kernel: services: add sys_irq

Add sys_irq section to docs

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

rruuaanng commented on include/zephyr/drivers/intc/irq_flags.h:
Can you add a sample `dts` file in the comment? The binding-API does not have full context, which confuses users (including me) :(

Like this
```
 *     intc1: interrupt-controller@... {
 *             compatible = "vnd,interrupt";
 *             #interrupt-cells = <2>;
 *     };
 *
 *     intc2: interrupt-controller@... {
 *             compatible = "vnd,interrupt";
 *             #interrupt-cells = <2>;
 *     };
 *
 *     n: node {
 *             interrupts-extended = <&intc1 3 2 15>,
 *                                   <&intc2 4 3 16>;
 *     };
```

bjarki-andreasen replied:
Added the following to the descriptions of the macros for getting IRQ flags from devicetree in `include/zephyr/drivers/intc/irq_flags.h`:
```
 * @details The encoding of the IRQ's flags is specific to each interrupt
 * controller, tied to its devicetree compatible, defined in
 * include/zephyr/drivers/intc/irq_flags.h
```

rruuaanng commented on doc/kernel/services/sys_irq.rst:
```diff
@@ -0,0 +1,58 @@
+Any and all configuration of INTLs is performed by SYS IRQ through the INTC device
+driver API.
+
+INTD (INTerrupt generating Device)
```
I think the abbreviation and the full name should be separated. Simply put, the whole process should be capitalized according to the first letter of the sentence, without deliberately showing the abbreviation.

For example
```
INTD (Interrupt generating device)
```

(Applicable to all changes.)

rruuaanng commented:
@bjarki-andreasen

ajf58 commented:
I'm trying to write an interrupt controller for the RP2350 to support the Hazard3 (RISC-V) cores. Currently, is it fair to assume this PR isn't going to land in the next 3 months?

bjarki-andreasen commented:
> I'm trying to write an interrupt controller for the RP2350 to support the Hazard3 (RISC-V) cores. Currently, is it fair to assume this PR isn't going to land in the next 3 months?

yes, I'm working full time on power management refactoring ATM :)
